internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher.go

```diff
@@ -93,52 +115,328 @@ func (c *gitHubWatcherController) Sync(ctx controllerlib.Context) error {
 	// same set of conditions.  Conditions depending on other conditions should get Status:  metav1.ConditionUnknown, or
 	// Status:  metav1.ConditionFalse, never be omitted.
-	conditions := []*metav1.Condition{
-		// we may opt to split this up into smaller validation functions.
-		// Each function should be responsible for validating one logical unit and setting one condition.
-		// c.validateGitHubAPI(),
-		// c.validateClaims(),
-		// c.validateAllowedAuthentication(),
-		// c.validateClient(),
+
+	hasErrorCondition := conditionsutil.HadErrorCondition(conditions)
+	updateStatusErr := c.updateStatus(ctx.Context, upstream, conditions)
+	if updateStatusErr != nil {
+		applicationErrors = append(applicationErrors, updateStatusErr)
+	}
+	if hasErrorCondition {
+		return nil, errors.NewAggregate(applicationErrors)
+	}
+
+	providerConfig := &upstreamgithub.ProviderConfig{
+		Name:               upstream.Name,
+		ResourceUID:        upstream.UID,
+		Host:               host,
+		GroupNameAttribute: groupNameAttribute,
+		UsernameAttribute:  usernameAttribute,
+		OAuth2Config: &oauth2.Config{
+			ClientID:     clientID,
+			ClientSecret: clientSecret,
+		},
+		AllowedOrganizations:    upstream.Spec.AllowAuthentication.Organizations.Allowed,
+		OrganizationLoginPolicy: policy,
+		AuthorizationURL:        fmt.Sprintf("%s/login/oauth/authorize", host),
+		HttpClient:              httpClient,
+	}
+	return providerConfig, errors.NewAggregate(applicationErrors)
+}
+
+func (c *gitHubWatcherController) validateHostAndTLSConfiguration(ctx context.Context, gitHubAPIConfig v1alpha1.GitHubAPIConfig) ([]*metav1.Condition, string, *http.Client, error) {
+	githubConnectionUnknown := &metav1.Condition{
+		Type:    GitHubConnectionValid,
+		Status:  metav1.ConditionUnknown,
+		Reason:  "UnableToValidate",
+		Message: "unable to validate; see other conditions for details",
+	}
+
+	tlsConfigurationUnknown := &metav1.Condition{
+		Type:    TLSConfigurationValid,
+		Status:  metav1.ConditionUnknown,
+		Reason:  "UnableToValidate",
+		Message: "unable to validate; see other conditions for details",
+	}
+
+	// Should not happen due to CRD defaulting
+	if gitHubAPIConfig.Host == nil || len(*gitHubAPIConfig.Host) < 1 {
+		return []*metav1.Condition{
+			{
+				Type:    HostValid,
+				Status:  metav1.ConditionFalse,
+				Reason:  "InvalidHost",
+				Message: `spec.githubAPI.host ("") is not valid`,
+			},
+			tlsConfigurationUnknown,
+			githubConnectionUnknown,
+		}, "", nil, nil
+	}
+
+	host := *gitHubAPIConfig.Host
+	hostPort, addressParseErr := endpointaddr.Parse(host, 443)
+	if addressParseErr != nil {
+		// addressParseErr is not recoverable. It requires user interaction, so do not return the error.
+		return []*metav1.Condition{
+			{
+				Type:    HostValid,
+				Status:  metav1.ConditionFalse,
+				Reason:  "InvalidHost",
+				Message: fmt.Sprintf("spec.githubAPI.host (%q) is not valid", host),
```

Comment:
Lets pass in the err to the `fmt.Sprintf("....")` so the user DOES get the specific error.  I don't think the user will likely go back to the CRD to infer a possible error.

Reply:
Example: https://github.com/vmware-tanzu/pinniped/blob/main/internal/controller/authenticator/jwtcachefiller/jwtcachefiller.go#L245